jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java

```diff
@@ -386,12 +420,21 @@ public static class ConcurrentEntry<E> implements Entry<E>
         // Other threads accessing must check the state field above first, so a good before/after
         // relationship exists to make a memory barrier.
         private E pooled;
+        // A weak ref to the pooled object; the hard ref gets nulled out when it is read
+        // and re-set when it is released, so if the pooled object is never released and gets
+        // garbage collected, this can be detected as this weak ref will become null.
+        private WeakReference<E> weakPooled;
```

Comment:
Note also that this in the case if a `ByteBufferPool`, this weak reference is to the `RetainableByteBuffer` and not to the `ByteBuffer` itself.
There is definitely a case of misuse, where the user discards the reference to the `RetainableByteBuffer`, but keeps a link to the `ByteBuffer` itself.     I think this is OK, as once they loose the reference to the `RetainableByteBuffer` they can never release and thus the buffer is effectively leaked.    We can remove it from the pool and create a new entry, which will allocate more memory even though the original buffer might still be referenced.   But this is why this event should be a WARNING and not a normal way to "release" a buffer.

Reply:
That's the gist of the idea behind this change. But it has one serious side-effect: we must guarantee that the code that uses this pool always keeps a hard ref to the pooled object.

This may not always be trivial and will require revisiting all our code that uses the concurrent pool: if the object acquired from the pool isn't immediately stored in a field, we will have to add some reachability fences like I did in `testLeakDetection`.